### ISSUE TITLE
Propagation of ManyToOne back to collections

### ISSUE THREAD

**Issue by stephenh** [mikro-orm/mikro-orm#307]
I'm writing some business logic / tests that creates / updates entities based on the `ManyToOne` side of a relation, and later looks / asserts against the `OneToMany` side of that same relation, but it is empty.

I.e.:

```
  it("should update posts", () => {
    const author = new Author();
    const post = new Post();
    post.author = author.toReference();
    expect(author.posts.getItems()).toContain(post);
  });
```

At first I thought this was a bug in `getItems()` thinking that the collection was initialized when really it was not. However, that is not the case, the collection is initialized (as empty since `Author` is a new entity), it's just not being kept up-to-date as the `post.author` field changes.

This means that to correctly have `author.posts.getItems()` to have the correct value, I have to: a) flush the author and post entities to the db, b) reload the author entity and c) init its collection.

This turns the unit of work / entities / collections into a leakier abstraction, as now I need to realize collections might return invalid values that don't respect the work-in-progress that exists in the rest of my unit of work. (I.e. I think ideally the unit of work is supposed to represent a unified / update-to-date view of currently-loaded entities, i.e. instead of re-fetching a row from the db everytime and missing "oh we've changed the author's name in this UoW so use that value").

So, I'd like to see this "we have some WIP and everyone sees that WIP consistently" behavior of the unit of work extended from primitives to these collections.

...all that said, it looks like the docs already highlight this propagation works if relationships are setup through the `Collection.add/remove` methods, I'd just like to see the same thing happen on the other side, so that my code doesn't have to be coupled to "to get the most consistent behavior out of Mikro, we always need to modify relationships via the `Collection` side".

**Comment by B4nan** [comment]
I don't think this is possible in JS (if I knew how to do this, it would be already there :]). Only way I could hook something to changing entity props is via Proxy object, but afaik that means you would have to work with the proxy instance, not with the underlying entity. 

The propagation works only on actions done on the `Collection` wrapper. 

> This means that to correctly have author.posts.getItems() to have the correct value, I have to: a) flush the author and post entities to the db, b) reload the author entity and c) init its collection.

You can always do it yourself:

```typescript
  it("should update posts", () => {
    const author = new Author();
    const post = new Post();
    post.author = author.toReference();
    author.posts.add(post); // manually
    post.setAuthor(author); // or via setter that does the very same thing
    expect(author.posts.getItems()).toContain(post);
  });
```

**Comment by stephenh** [comment]
> I don't think this is possible in JS 

Well, yes & no; you did mention the proxy option (which I agree can be useful, but is good to avoid if possible), but fwiw this is generally what I was imagining with the `.set` method of `IdentifiedReference` would do (although probably as a new `Collection`-ish type and not `IdentifiedReference` itself).

I.e. just like `Collection` is always instantiated during entity construction, and so available to `.add`/`.remove` methods, if `ManyToOne`s were modeled like:

```
@ManyToOne
readonly author = new SomethingLikeCollection(this);
```

And then getting/setting was done via `post.author.get()` / `post.author.set(...)`.

Granted, this doesn't support the "using `populate` to pre-fill the `post.author` and then accessing it like a regular / always-available synchronous property / primitive access".

But, admittedly you know my bias on this :-), accessing `ManyToOne`'s is not really like accessing primitive fields, and it is much closer to how `OneToMany`'s collections are modeled, so I think in terms of consistency, this proposed `readonly author = new SomethingLikeCollection(this)` is more consistent, plus provides a very obvious hook for: a) propagation to the collection on the other side and b) runtime checks to ensure loaded-ness, which currently are just silent runtime errors (...iirc, I don't use this approach in mikro so I'm guessing a little bit on its exact semantics).

wdyt? Is it maybe likely to see something like `author = new SomethingLikeCollection` in a sometime-soon-ish v3 point release? v4?

 > You can always do it yourself:

True, although that seems somewhat error prone to have to leave to unit tests and code reviews to remind the team of the right approach...

Although looking at your example code, I wonder if property getters/setters could do something like that, i.e.:

```
class Post {
  // only used by mikro
  @ManyToOne(...)
  private _author: IdentifiedRef...

  // used by the application code
  get author(): Author {
    return this._author;
  }

  set author(author: Author) {
    this._author = author
    author.posts.add(this);
  }
}
 ```

Or some variation on that. I might play around with that. It would be tedious to write by hand, but I'm generating all of our fields/annotations off the database schema (not using mikro's entity generator b/c I wanted to play with things like IdentifiedRefs/etc.)... 

I suppose if that approaches works out, the `author = new SomethingLikeCollection` approach would not be necessary, although at this point I still generally prefer that I think, granted b/c it fits my particular use case/pain point.

**Comment by B4nan** [comment]
Hmm maybe I could be able to do this trick automatically. I thought I won't be able to redefine entity properties, but the truth is that they do not exist during runtime (on the prototype level), so I can use `Object.defineProperty()` to set up the `get/set` on the prototype during discovery. 

https://www.typescriptlang.org/play/?target=99#code/MYGwhgzhAEBiD29oG8BQ0PQEZgE4C5oA7AVwFssBTXaAXmgGYAmARgG5UBfVVAeSwBWlYABcAdABNKAMwCWRSgAVc8AA7URATwAUCeGNUqR8LeoA00AOQB9axLAiwli8mgA3MCBKVCyTtE4ASj5BYXEpOQVlNQ0dPQMjE01zKxxcZxR0TABzShFtQMzMYuhgeCIIeBBKMRB4bO1LXJEMkQALWQgxW3tHMTTAjhLMXDySXCJods7uuwcwfrwhzE4zLIwIPO0PEEIIEVx5bMK0YYwyiqqauobLTZaLHcH1kumunvnFmnod5YxVrjBVAXfbQaR0YiUADucEQBQ40i+CK+EJYTAYyLSHFQQA

**Comment by B4nan** [comment]
Looks like it is working, but it have one downside. Even when I define the property as `enumerable`, it is not present when `console.log`ing the entity. That could be solved by custom `util.inspect` method on the prototype, but if users would try to iterate the properties, it would also not work as `Object.keys(entity)` also ignores properties defined on the prototype. Because of this, there are quite a lot of broken tests, but that can be definitely fixed. 

We could have this conditionally (not sure about the default). The thing is that it feels a bit strange, one could would work with the property like usual, it would be present when `console.log`ing, but not trying to iterate thru the props. Personally I never do this, and if I would want to do something similar, I would first call `entity.toObject()` (where the issue would not be present). 

WDYT?

https://www.typescriptlang.org/play/?target=99&ssl=23&ssc=68&pln=23&pc=66#code/MYGwhgzhAEBiD29oG8BQ0PQEZgE4C5oA7AVwFssBTXaAXmgGYAmARgG51McAvQiAF1wBLIgHM60AOT9KAyRwC+qVAHksAK0rB+AOgAmlAGYjKABVzwADtX4BPABQJ4Oyxf7w71gDRSA+r70wfjBJH2RoADcwEBJKQmQFaAUASlUNLV0DYyIzC2tcO0dEFzcPW28pHFxQlE4MUUp+e2TazDboYHgiCHgQSh0QeFF7SQb+Gv4ACyEIHX9A4J0q5I52zFxGklwiaCmZuYCgsCW8VcwFLzroCEb7KJA+QRFRFrQ1jE7u3v7B4ckb8Y+e4rK5tPazeZHE40ej3M4YC5XSikMjUMBYPqEQSxS4pDioT4CaCGCQ5ADucEQzQ4hmhNOhEhYTAY9KqHEJ3wGQ3sak02h0AGtKLYIPZDMkfLyMoLhaKpfyxuYPGVrCpDGLkpqOEA

also tagging @lookfirst for another view

**Comment by B4nan** [comment]
I think I found a workaround. First, I create a setter on the prototype. This setter will then create new getter and setter on the actual instance when called for the first time. This way everything works ok. 

https://www.typescriptlang.org/play/?target=99&ssl=46&ssc=7&pln=46&pc=1#code/MYGwhgzhAEBiD29oG8BQ0PQEZgE4C5oA7AVwFssBTXaAXmgGYAmARgG51McAvQiAF1wBLIgHM60AOT9KAyR0zRg8IgNwlg-eLgAUAShSdFGZavghKAOhDxROydqHjN2yXoWYAvqm+oAZiREmkIq0AAmlMq4YDI6uJR+hPBYAFaR-AA00AAOuPBaEoIklAZoiqYQ5lY2dpIRUTGUbhxG0EJ+0PYA+l1hMWCSbUTQ8X6lreUqlRbWtvZgIPFgYQCe4ZHajWHNE5jx-CS4RB4Yvortnbn58OPGmADyqemWEX4ilAAKednU-CtxCSykhwuEkWTKd2MEEo-B0ADcFnxBCJRLdIej6ptYvwABZCCBZPwLaHuXbojC4-GWEESBEgE7kzwZMnQTykxSeaCUEDQwzox5pTQvBLvL7wH64P4AvxAnp9fgDcHQOnFQjITlshmKAXPV6i76-f6jIEgsF88kYUQw-Tmi2TMwzGr2K38M2UiCWOX9al4dl2xT7Q7Dd2e3rekFayFMlmKaGwulI4RiNH+kxTKqzWpxs10v2p6Ahr0Kn00eh0yN3aP+yikMjUMBYCyEIqUZnozWtby+THRWIIeCWK5aP4-UmoCr8aAdehESgAdzgiH0HD8JZXJYkLCYDHXNPoABYAKwANl3eA4FQzTp1QoA1pQVhAdGMsjf+JZ74+dG-LC6xcOVh+e4-GfPQwI4IA

**Comment by stephenh** [comment]
@B4nan ah wow, that double-`defineProperty` approach is pretty nifty.

Definitely agree the property staying in `Object.keys` would be important, so your double-`defineProperty` approach would definitely be the way to go if staying with the current `parent: Parent` (vs. `parent: SomethingCollectionLike<Parent>`).

I'm tempted to say I still like the consistency and simplicity of the wrapper `SomethingCollectionLike`, that would have both async (for when you don't know if its loaded) and sync (when you do know its loaded) methods on it. It seems closer to modeling reality, in that this is not literally a `Parent` in memory, but instead is a reference that might/might not be in memory, through which you can `get`/`load`/etc. a `Parent`...

Also, if you did introduce a "must be loaded" check to `child.parent` accesses via a `defineProperty` getter, then I think that would mean use cases like `await child.parent.init()` would not work?

Interesting approach though...

**Comment by B4nan** [comment]
> Also, if you did introduce a "must be loaded" check to child.parent accesses via a defineProperty getter, then I think that would mean use cases like await child.parent.init() would not work?

Not sure what you mean, getter would always return, only setter in this case would propagate the change when the inverse side (1:m collection) is initialized (as you cannot modify collection that is not).

Already got this implemented and it works fine, with a downside of `console.log(entity)` not printing getter properties by default. That can be solved by `inspect.custom`, but it's kinda tricky to implement it correctly (computing correct padding for nested objects). Now I am trying to use this also for 1:1, but there few failing tests. 

Here is the code for patching the prototype:

```typescript
  private static defineReferenceProperties<T extends AnyEntity<T>>(meta: EntityMetadata<T>) {
    Object
      .values<EntityProperty>(meta.properties)
      .filter(prop => [ReferenceType.ONE_TO_ONE, ReferenceType.MANY_TO_ONE].includes(prop.reference) && (prop.inversedBy || prop.mappedBy))
      .forEach(prop => {
        Object.defineProperty(meta.prototype, prop.name, {
          set(val: AnyEntity) {
            if (!('__data' in this)) {
              Object.defineProperty(this, '__data', { value: {} });
            }

            EntityHelper.defineReferenceProperty(prop, this, val);
          },
        });
      });
  }

  private static defineReferenceProperty<T extends AnyEntity<T>>(prop: EntityProperty<T>, ref: T, val: AnyEntity) {
    Object.defineProperty(ref, prop.name, {
      get() {
        return this.__data[prop.name];
      },
      set(val: AnyEntity) {
        this.__data[prop.name] = val;
        const inverse = val && val[prop.inversedBy || prop.mappedBy];

        if (prop.reference === ReferenceType.MANY_TO_ONE && inverse && inverse.isInitialized()) {
          (inverse as Collection<T>).add(this);
        }

        if (prop.reference === ReferenceType.ONE_TO_ONE && val && wrap(val).isInitialized() && inverse !== this) {
          val[prop.inversedBy || prop.mappedBy] = this;
        }
      },
      enumerable: true,
      configurable: true,
    });
    ref[prop.name] = val as T[string & keyof T];
  }
```

**Comment by B4nan** [comment]
So I finally cracked all of the problems that this change uncovered. They were all there but not visible as one usually do not make the relation fully propagated in the entity graph. 

As part of the refactorings I had to introduce commit order calculator, pospone m:n collection queries, also there were bugs in EntityFactory. 

Will try to split the work into multiple commits now so all the changes are visible in the history, but finally all tests are now passing.
